### DISCUSSION TITLE
Why doesn't AddCors or AddJwtBearer have support for getting a registered DI/IoC item?

### DISCUSSION THREAD

**Discussion by PureKrome** [dotnet/aspnetcore#57750] [Q&A] [answered]
Hi folks,

So we have `AddCors` and `AddJwtBearer` in our code. both of these have some specific code to set values which are based on Configuration.

for example:

![image](https://github.com/user-attachments/assets/9e6f1348-eb6a-4b64-96dc-32787e0b9be3)

The allowed origins are defined in configuration files/environment variables.

Likewise with setting up authentication's AddJwtBearer .. we have this:

![image](https://github.com/user-attachments/assets/02f96582-8679-4c50-b6fd-14cf5c81a474)

again, some settings via configuration.

I couldn't find any way to grab this from the existing DI/IoC if we have previously registered it. My guess is that the above code is still being handled/called _during registration_ of all our services.

So could this be added to the code? Would this mean that cors and jwtsettings would be setup upon first instansiation / request of the instance/data?

It's like I wish to load all configuration data _before_ configuring any services .. but this problem is that i load configuration settings during service configuration (into `IOptions<SomeConfigSection>`).

so confused :(

**Comment by pinkfloydx33** [comment, marked as answer]
Those are just the _initial_ setting values, you can still configure the values after the calls to `AddCors`/`AddJwtBearer` but before the service provider is built. Just register `IConfigureOptions<CorsOptions>` or `IConfigureOptions<JwtBearerOptions>` and inject your configuration dependencies into those. You can then use `Configure` (or `PostConfigure` with `IPostConfigureOptions<>`) to complete the configuration using your dynamic values

**Reply by PureKrome** [reply]
Hi @pinkfloydx33 - sorry for the late reply. I don't understand this :(

can you provide some code example?

**Comment by pinkfloydx33** [comment]
Something like the below (typing from mobile so may not be exactly correct). You can inject any DI service, including `IConfiguration` or `IOptions<>`

```csharp 
public sealed class ConfigureMyJwtOptions(ISomeService myService) : IPostConfigureOptions<JwtBearerOptions>
{
    public void PostConfigure(string name, JwtBearerOptions opts) 
    {
        // limit to the specific JWT  registration since you can technically have multiples by 
        // checking the name matches the one you used in the call to AddJwtBearer. 
        // If you don't specify one it uses the name from JwtBearerDefaults.
        // Or you can ignore the name parameter and apply to all JetBearerOptions
        // if (name == JwtBearerDefaults.AuthenticationScheme) {...} 
        opts.TokenValidationParameters.ValidIssuer = myService.GetIssuer();
    } 
}
// in startup 
services.ConfigureOptions<ConfigureMyJwtOptions>();
```

Or implement `IConfigureOptions<>` and it's `Configure` method. Registration order matters; since `AddJwtBearer` also registers options configuration you'd either need to call `ConfigureOptions` before `AddJwtBearer` to ensure your configuration takes precedence, or you can use `IPostConfigureOptions` like my example which will run after all other options setup

**Reply by PureKrome** [reply]
wait what? There's `IPostConfigureOptions` ? What magic is this and why have I never seen this?

Also .. got a message (in another Slack) from @kevinchalet who has also dropped some bombs about this... brb

_<SpongeBob narrator voice>**A few moments later**</SpongeBob narrator voice>_

Oh ok - so it's like using the custom configuring the Options class.

E.g. `AddOptions<JwtBearerOptions>().Configure<IServiceProvider>(static (options, serviceProvider) => {} );`

wow.

```
services
    .AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddJwtBearer();

services.AddOptions<JwtBearerOptions>(JwtBearerDefaults.AuthenticationScheme)
    .Configure<IServiceProvider>(static (options, serviceProvider) =>
    {
        var jwtSettings = serviceProvider.GetRequiredService<IOptions<JwtSettings>>().Value;

        // Stop the 'sub' claim from getting renamed to 'nameIdentifier' claim
        options.MapInboundClaims = false;

        var issuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(jwtSettings.IssuerSigningKey));

        options.TokenValidationParameters = new TokenValidationParameters
        {
            ValidateIssuer = true,
            ValidateAudience = true,
            ValidateLifetime = true,
            ValidateIssuerSigningKey = true,
            ValidAudiences = [jwtSettings.PublicAudience, jwtSettings.AdminAudience],
            ValidIssuer = jwtSettings.Issuer,
            IssuerSigningKey = issuerSigningKey,
            RoleClaimType = CommonConstants.CLAIM_ROLE
        };
    });

services.AddCors();
services.AddOptions<CorsOptions>()
    .Configure<IServiceProvider>(static (options, serviceProvider) =>
    {
        var corsSettings = serviceProvider.GetRequiredService<IOptions<CorsSettings>>().Value;

        options.AddDefaultPolicy(policy =>
        {
            policy
                .AllowAnyHeader()
                .AllowAnyMethod()
                .AllowCredentials()
                .SetIsOriginAllowedToAllowWildcardSubdomains()
                .WithOrigins(corsSettings.Origins)
                .WithExposedHeaders("Location");
        });
    });

return services;

```

massive hat tip also to @kevinchalet for his help on this.

next .. need to confirm that this is documented for newbies like me .... 🔎

**Reply by pinkfloydx33** [reply]
The way you registered the options there (with OptionsBuilder) is actually implemented under the hood with `IConfigureOptions<>` implementations (actually/technically `IConfigureNamedOptions<>` in the first case which is just a specialization and mostly an internal thing you'd not use directly) 

Also injecting a service provider is an antipattern, especially here where you can just inject the right thing...

```csharp
services.AddOptions<CorsOptions>()
    .Configure<IOptions<CorsSettings>>(static (options, corsOpts) =>
    {
        var corsSettings = corsOpts.Value;
        // the rest
    });
```

**Reply by PureKrome** [reply]
oh even better!

**Reply by pinkfloydx33** [reply]
Personally I avoid this route and will almost always use `IConfigureOptions`/`IPostConfigureOptions` implementations, especially since those then because very easily unit testable.

**Reply by PureKrome** [reply]
Can you provide an example of using `IConfigureOptions` please? to compare to the `.Configure<...>` example.

**Reply by pinkfloydx33** [reply]
I did above.... But here's another one 

```
public sealed class ConfigureMyCorsOptions(IOptions<CorsSettings> corsOpts) : IConfigureOptions<CorsOptions>
{
    public void Configure(CorsOptions opts) 
    {
        var corsSettings = corsOpts.Value;
        opts.AddDefaultPolicy(policy =>
        {
            policy
                .AllowAnyHeader()
                .AllowAnyMethod()
                .AllowCredentials()
                .SetIsOriginAllowedToAllowWildcardSubdomains()
                .WithOrigins(corsSettings.Origins)
                .WithExposedHeaders("Location");
        });
    } 
}

// in startup 
// registers all IConfigureOptions, IPostConfigureOptions and IValidateOptions in the target class 
services.ConfigureOptions<ConfigureMyCorsOptions>(); 

// or directly
services.AddSingleton<IConfigureOptions<CorsOptions>, ConfigureMyCorsOptions>();
```

There's definitely a bit more ceremony since you need to define a whole class. But now you can write tests to verify you're configuration logic is working correctly (without needing to startup an entire copy of the application)

**Reply by pinkfloydx33** [reply]
You _could_ even define a single class that Configures CORS and Jwt options, and maybe validates them too

```csharp
public sealed class MyConfigurator(IOptions<CorsSettings> corsOpts, IOptions<JwtSettings> jwtOpts) :
  IConfigureOptions<CorsOptions>,
  IPostConfigureOptions<JwtBearerOptions>,
  IValidateOptions<JwtBearerOptions>
{
  public void Configure(CorsOptions opts)
  {
    var corsSettings = corsOpts.Value;
    // configure opts
  }

  public void PostConfigure(string name, JwtBearerOptions opts)
  {
    if (name != JwtBearerDefaults.AuthenticationScheme)
      return;
    
    var jwtSettings = jwtOpts.Value;
    // configure opts 
  }

  public ValidateOptionsResult Validate(string name, JwtBearerOptions opts)
  {
    if (name != JwtBearerDefaults.AuthenticationScheme) 
      return ValidateOptionsResult.Skip;

    if (string.IsNullOrWhiteSpace(opts.TokenValidationParameters?.ValidIssuer)) 
      return ValidateOptionsResult.Fail("valid issuer is missing");

    return ValidateOptionsResult.Success;
  } 
} 
```

And then you can register them all at once:

```csharp 
services.ConfigureOptions<MyConfigurator>();
```

Though I wouldn't personally combine configuration like this into one class unless they were all for the same options type... This is just for demo purposes (and again, it was all typed from mobile, so hopefully there's no typos)
